src/validation/validateProperties.ts

```diff
@@ -0,0 +1,167 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: MIT
+ */
+
+import {
+    ArrayASTNode,
+    ASTNode,
+    Diagnostic,
+    DiagnosticSeverity,
+    ObjectASTNode,
+    PropertyASTNode,
+    Range,
+    TextDocument,
+} from 'vscode-json-languageservice';
+
+import {
+    findPropChildByName,
+    isArrayNode,
+    isObjectNode,
+} from '../utils/astUtilityFunctions'
+
+import { MESSAGES } from '../constants/diagnosticStrings'
+
+import schema from './validationSchema'
+
+function isObject(obj: any): obj is Object {
+    return obj === Object(obj);
+  }
+
+function getDiagnosticsForNode(
+    rootNode: ASTNode,
+    document: TextDocument,
+    schemaPart: Object,
+    schemaSets: Object
+): Diagnostic[] {
+    let diagnostics: Diagnostic[] = []
+
+    const arraySchema = schemaPart['Fn:ArrayOf']
+    const oneOfSchema = schemaPart['Fn:OneOf']
+    const valueOfSchema = schemaPart['Fn:ValueOf']
+
+    // Fn:ArrayOf
+    // if it contains Fn:ArrayOf property all the other values will be ignored
+    if (typeof arraySchema === 'string' && isArrayNode(rootNode)) {
+        const newSchemaPart: unknown = schemaSets[arraySchema]
+
+        if (isObject(newSchemaPart)) {
+            rootNode.items.forEach(itemNode => {
+                if (isObjectNode(itemNode)) {
+                    diagnostics = diagnostics.concat(getDiagnosticsForNode(itemNode, document, newSchemaPart, schemaSets))
+                }
+            })
+        }
+
+        return diagnostics
+
+    // Fn:OneOf
+    // TODO: Allow for both OneOf and ValueOf at the same time
+    } else if (typeof oneOfSchema === 'string' && isObjectNode(rootNode)) {
+        const oneOfSet: unknown = schemaSets[oneOfSchema]
+        const itemsOfOneOfSet: { propNode: PropertyASTNode, schemaValue: unknown }[] = []
+
+        rootNode.properties.forEach(prop => {
+            const propName = prop.keyNode.value
+            const inOneOfSet = oneOfSet[propName]
+
+            // If the property is in set referenced by Fn:OneOfSet
+            if (inOneOfSet) {
+                itemsOfOneOfSet.push({ propNode: prop, schemaValue: inOneOfSet})
+            // If the property is neither in the set nor in the schema props
```

Comment:
I'm confused what "set" and "schema props" refer to here. I think a descriptive variable name to compare against should be self-explanatory. `isValidProperty` was suggested above